File: test_interpolation.pl
#!/usr/bin/env perl

# Test complex string interpolation edge cases

# Basic interpolation (should work)
"Hello $name";
"Array: @items";
"Hash element: $hash{key}";

# Complex interpolation cases
"Complex: @{[ $x + $y ]}";
"Deref: @{$arrayref}";
"Hash deref: %{$hashref}";
"Code eval: @{[ func($x) ]}";
"Nested: @{[ $arr->[$i] ]}";

# Multiple interpolations
"Name: $first $last";
"List: @items, count: $#items";

# Escaped interpolation
"Literal \$dollar";
"Literal \@at";

# Special variables
"PID: $$";
"Error: $!";
"Match: $1, $2";

# Method calls in interpolation
"Result: $obj->method()";
"Chain: $obj->foo()->bar()";

# Array/hash slices in interpolation
"Slice: @array[0..2]";
"Hash slice: @hash{'a', 'b'}";

# Mixed quotes
qq{Hello $name};
qq[Array @items];
qq|Path: $ENV{PATH}|;

File: crates/perl-parser/examples/test_interpolation.rs
//! Test string interpolation
use perl_parser::Parser;

fn main() {
    let tests = vec![
        // Simple strings (no interpolation)
        r#"'hello'"#,
        r#"'hello world'"#,
        
        // Double-quoted strings (interpolation)
        r#""hello""#,
        r#""hello world""#,
        r#""hello $name""#,
        r#""hello ${name}""#,
        r#""array: @array""#,
        r#""hash: %hash""#,
        
        // Escape sequences
        r#""hello\n""#,
        r#""hello\tworld""#,
        r#""quote: \"""#,
        
        // Complex interpolation
        r#""$hash{key}""#,
        r#""$array[0]""#,
        r#""$obj->method()""#,
        r#""${$ref}""#,
        
        // Mixed
        r#""Hello $name, you have $count items""#,
        r#""Path: $ENV{PATH}""#,
    ];
    
    for test in tests {
        println!("\nTesting: {}", test);
        let mut parser = Parser::new(test);
        match parser.parse() {
            Ok(ast) => {
                println!("✅ Success!");
                println!("   S-expr: {}", ast.to_sexp());
            }
            Err(e) => {
                println!("❌ Error: {}", e);
            }
        }
    }
}